File: config-auth-core-basic-kt/assets/auth-core-basic-template-kt/src/main/kotlin/com/example/auth/user/usecase/LoginUseCase.kt
package com.example.auth.user.usecase

import com.example.auth.password.provider.PasswordCryptoProvider
import com.example.auth.user.dto.UserDto
import com.example.auth.user.query.FindPasswordHashQuery
import com.example.auth.user.query.FindUserByEmailQuery

class LoginUseCase(
    private val findUserByEmail: FindUserByEmailQuery,
    private val findPasswordHash: FindPasswordHashQuery,
    private val crypto: PasswordCryptoProvider
) {
    suspend fun execute(email: String, plainPassword: String): Result<UserDto> {
        val user = findUserByEmail.findByEmail(email)
            ?: return Result.failure(IllegalArgumentException("INVALID_CREDENTIALS"))
        val hash = findPasswordHash.findPasswordHash(user.id)
            ?: return Result.failure(IllegalArgumentException("INVALID_CREDENTIALS"))
        if (!crypto.compare(plainPassword, hash)) {
            return Result.failure(IllegalArgumentException("INVALID_CREDENTIALS"))
        }
        return Result.success(user)
    }
}

File: config-auth-core-basic/assets/auth-core-basic-template/src/user/usecase/login.usecase.ts
import { FindPasswordHashQuery, UserRepository } from '../provider';
import { PasswordErrors, PasswordCryptoProvider } from '../../password';
import { Result, UseCase } from '__SHARED_PACKAGE_NAME__';
import { UserProps } from '../model';

export interface LoginIn {
  email: string;
  password: string;
}

export interface LoginOut extends UserProps {}

export class LoginUseCase implements UseCase<LoginIn, LoginOut> {
  constructor(
    private readonly repo: UserRepository,
    private readonly findPassHash: FindPasswordHashQuery,
    private readonly passwordCryptoProvider: PasswordCryptoProvider,
  ) {}

  async execute(input: LoginIn): Promise<Result<LoginOut>> {
    return Result.try(async () => {
      const tryFindUser = await this.repo.findByEmail(input.email);
      if (tryFindUser.isFailure) return Result.fail(tryFindUser.errors);

      const tryFindPass = await this.findPassHash.execute(tryFindUser.instance.id);
      if (tryFindPass.isFailure) return Result.fail(tryFindPass.errors);

      const isSamePass = await this.passwordCryptoProvider.compare(input.password, tryFindPass.instance.hash);
      if (!isSamePass) return Result.fail(PasswordErrors.MISMATCH);

      return tryFindUser.instance.props;
    });
  }
}
